src/Adapter/MSTest.TestAdapter/Extensions/MethodInfoExtensions.cs

```diff
@@ -188,4 +194,103 @@ internal static void InvokeAsSynchronousTask(this MethodInfo methodInfo, object?
             valueTask.GetAwaiter().GetResult();
         }
     }
+
+    // Scenarios to test:
+    //
+    // [DataRow(null, "Hello")]
+    // [DataRow("Hello", null)]
+    // public void TestMethod<T>(T t1, T t2) { }
+    //
+    // [DataRow(0, "Hello")]
+    // public void TestMethod<T1, T2>(T2 p0, T1, p1) { }
+    private static MethodInfo ConstructGenericMethod(MethodInfo methodInfo, object?[]? arguments)
+    {
+        DebugEx.Assert(methodInfo.IsGenericMethod, "ConstructGenericMethod should only be called for a generic method.");
+
+        if (arguments is null)
+        {
+            // An example where this could happen is:
+            // [TestMethod]
+            // public void MyTestMethod<T>() { }
+
+            // TODO: Localize.
+            throw new TestFailedException(ObjectModel.UnitTestOutcome.Error, $"The generic test method '{methodInfo.Name}' doesn't have arguments, so the generic parameter cannot be inferred.");
+        }
+
+        Type[] genericDefinitions = methodInfo.GetGenericArguments();
+        var map = new (Type GenericDefinition, Type? Substitution)[genericDefinitions.Length];
+        for (int i = 0; i < map.Length; i++)
+        {
+            map[i] = (genericDefinitions[i], null);
+        }
+
+        ParameterInfo[] parameters = methodInfo.GetParameters();
+        for (int i = 0; i < parameters.Length; i++)
+        {
+            Type parameterType = parameters[i].ParameterType;
+            if (parameterType.IsGenericMethodParameter())
+            {
+                if (arguments[i] is null)
+                {
+                    continue;
+                }
+
+                Type substitution = arguments[i]!/*Very strange nullability warning*/.GetType();
+                int mapIndexForParameter = GetMapIndexForParameterType(parameterType, map);
+                Type? existingSubstitution = map[mapIndexForParameter].Substitution;
+
+                if (existingSubstitution is null || substitution.IsAssignableFrom(existingSubstitution))
+                {
+                    map[mapIndexForParameter] = (parameterType, substitution);
+                }
+                else if (existingSubstitution.IsAssignableFrom(substitution))
+                {
+                    // Do nothing. We already have a good existing substitution.
+                }
+                else
+                {
+                    // TODO: Localize.
+                    throw new InvalidOperationException($"Found two conflicting types for generic parameter '{parameterType.Name}'. The conflicting types are '{existingSubstitution}' and '{substitution}'.");
+                }
+            }
+        }
+
+        for (int i = 0; i < map.Length; i++)
+        {
+            // TODO: Better to throw? or tolerate and transform to typeof(object)?
```

Comment:
Can you explain the possible scenario we would end up here?

Reply:
```csharp
[DataRow(null)]
public void TestMethod<T>(T t) { }
```

Reply:
Ok then I am fine either way. I would probably go for `object` because it's not really "wrong" to do that.